### PR TITLE
feat(STONEINTG-493): split pipeline controller to 2 new controller

### PR DESCRIPTION
**Pull request by hongweiliu17**
* split pipeline controller to build and test pipeline controller
* move func PrepareSnapshot() to gitops pkg and refactor it
* rename the pipelinerun variable in unittest to buildPipelineRun
  or integrationPipelineRun

Signed-off-by: Hongwei Liu <hongliu@redhat.com>

### CODE REVIEW

File: controllers/testpipeline/testpipeline_controller.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions andF
 limitations under the License.
 */
 
-package pipeline
+package testpipeline
```
Comment (14rcole) [inline controllers/testpipeline/testpipeline_controller.go:17]: Can we call this `integrationpipeline` instead?  It lines up with the name of our service, but since there are also tests in the build pipeline it draws a clear line regarding what's being tested.

Comment (hongweiliu17) [reply controllers/testpipeline/testpipeline_controller.go:17]: Oh, it was changed back to `pipeline` when I rebase the code. Thanks!

Comment (hongweiliu17) [reply controllers/testpipeline/testpipeline_controller.go]: Sorry. I misunderstood your comment. Updated. Thanks!

File: controllers/buildpipeline/buildpipeline_controller_test.go [hunk 2]
```diff
@@ -41,14 +41,14 @@ import (
 
 var _ = Describe("PipelineController", func() {
 	var (
-		manager            ctrl.Manager
-		pipelineReconciler *Reconciler
-		scheme             runtime.Scheme
-		req                ctrl.Request
-		successfulTaskRun  *tektonv1beta1.TaskRun
-		testpipelineRun    *tektonv1beta1.PipelineRun
```
Comment (MartinBasti) [inline controllers/buildpipeline/buildpipeline_controller_test.go:49]: Why is this change needed?

File: controllers/integrationpipeline/integrationpipeline_controller.go [hunk 3]
```diff
@@ -133,13 +118,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 // AdapterInterface is an interface defining all the operations that should be defined in an Integration adapter.
 type AdapterInterface interface {
-	EnsureSnapshotExists() (controller.OperationResult, error)
 	EnsureSnapshotPassedAllTests() (controller.OperationResult, error)
 	EnsureStatusReported() (controller.OperationResult, error)
 	EnsureEphemeralEnvironmentsCleanedUp() (controller.OperationResult, error)
 }
 
-// SetupController creates a new Integration controller and adds it to the Manager.
+// SetupController creates a new Integration reconciler and adds it to the Manager.
```
Comment (MartinBasti) [inline controllers/integrationpipeline/integrationpipeline_controller.go:126]: https://github.com/redhat-appstudio/integration-service/pull/254/files in this PR, reconciler was replaced by controller, here you are doing it in the opposite way.

Also why do you have to do this extra change in this already big PR?

File: controllers/integrationpipeline/integrationpipeline_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
+/*
+Copyright 2022.
```
Comment (MartinBasti) [inline controllers/integrationpipeline/integrationpipeline_controller_test.go:2]: 2023

File: controllers/buildpipeline/buildpipeline_controller_test.go (hunk 2 above)
Comment (MartinBasti) [reply controllers/buildpipeline/buildpipeline_controller_test.go:49]: I mean renaming `test*` to `integration*`

File: controllers/buildpipeline/buildpipeline_controller.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+// SetupController creates a new Integration reconciler and adds it to the Manager.
+func SetupController(manager ctrl.Manager, log *logr.Logger) error {
+	return setupControllerWithManager(manager, NewIntegrationReconciler(manager.GetClient(), log, manager.GetScheme()))
+}
+
+// setupCache indexes fields for each of the resources used in the pipeline adapter in those cases where filtering by
+// field is required.
+func setupCache(mgr ctrl.Manager) error {
+	if err := cache.SetupApplicationComponentCache(mgr); err != nil {
```
Comment (dirgim) [inline controllers/buildpipeline/buildpipeline_controller.go:127]: Since we already set up the same cache in the integrationpipeline controller [here](Image repository where the built image was pushed) (temporary link), this causes a conflict.
Basically, we cannot set up the same type of cache in multiple places.

File: controllers/integrationpipeline/integrationpipeline_controller.go [hunk 6]
```diff
@@ -77,20 +77,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	pipelineRun := &tektonv1beta1.PipelineRun{}
 	err := r.Get(ctx, req.NamespacedName, pipelineRun)
 	if err != nil {
-		logger.Error(err, "Failed to get pipelineRun for", "req", req.NamespacedName)
+		logger.Error(err, "Failed to get test pipelineRun for", "req", req.NamespacedName)
```
[the remark this answers is not in the document]
Comment (hongweiliu17) [reply controllers/integrationpipeline/integrationpipeline_controller.go:80]: Yes, thanks!

File: controllers/buildpipeline/buildpipeline_controller.go (hunk 5 above)
Comment (hongweiliu17) [reply controllers/buildpipeline/buildpipeline_controller.go:127]: Thanks! Let me update it

File: controllers/integrationpipeline/integrationpipeline_controller.go (hunk 3 above)
Comment (MartinBasti) [reply controllers/integrationpipeline/integrationpipeline_controller.go:126]: IMO you answered to wrong thread, but I got it

File: controllers/buildpipeline/buildpipeline_controller_test.go (hunk 2 above)
Comment (hongweiliu17) [reply controllers/buildpipeline/buildpipeline_controller_test.go:49]: One is build pipelinerun, another is integration test pipelinerun, if we still use the old name, it looks weird and confusing, so I would change their name.

Comment (MartinBasti) [reply controllers/buildpipeline/buildpipeline_controller_test.go:49]: okay

File: controllers/integrationpipeline/integrationpipeline_controller.go (hunk 3 above)
Comment (hongweiliu17) [reply controllers/integrationpipeline/integrationpipeline_controller.go]: Don't know why git consider this `integrationpipeline` as a renamed folder of `pipeline`.

File: controllers/buildpipeline/buildpipeline_controller.go (hunk 5 above)
Comment (dheerajodha) [reply controllers/buildpipeline/buildpipeline_controller.go]: Even after pushing this change, I saw the CI fail with that same bug, am I right?
If yes, what was/were the other breaking change(s)?

File: controllers/buildpipeline/buildpipeline_controller_test.go [hunk 7]
```diff
@@ -41,14 +41,14 @@ import (
...
-		hasApp             *applicationapiv1alpha1.Application
-		hasComp            *applicationapiv1alpha1.Component
+		manager                ctrl.Manager
+		pipelineReconciler     *Reconciler
+		scheme                 runtime.Scheme
+		req                    ctrl.Request
+		successfulTaskRun      *tektonv1beta1.TaskRun
+		integrationpipelineRun *tektonv1beta1.PipelineRun
```
Comment (dirgim) [inline controllers/buildpipeline/buildpipeline_controller_test.go:49]: Would it make more sense to name this as `buildPipelineRun` since its type is `build`?

File: controllers/integrationpipeline/integrationpipeline_controller.go (hunk 3 above)
Comment (MartinBasti) [reply controllers/integrationpipeline/integrationpipeline_controller.go]: becuase marking it as moved instead of deleted results into less diff lines

File: loader/integration.go [hunk 8]
```diff
@@ -55,3 +60,93 @@ func GetLatestPipelineRunForSnapshotAndScenario(adapterClient client.Client, ctx
 
 	return nil, err
 }
+
+// prepareSnapshot prepares the Snapshot for a given application and the updated component (if any).
+// In case the Snapshot can't be created, an error will be returned.
+func PrepareSnapshot(adapterClient client.Client, ctx context.Context, loader ObjectLoader, application *applicationapiv1alpha1.Application, component *applicationapiv1alpha1.Component, newContainerImage string, newComponentSource *applicationapiv1alpha1.ComponentSource) (*applicationapiv1alpha1.Snapshot, error) {
```
Comment (dirgim) [inline loader/integration.go:66]: I'm not sure if `PrepareSnapshot` should go under the `loader` package, maybe [gitops/snapshot.go](https://github.com/redhat-appstudio/integration-service/blob/main/gitops/snapshot.go) would make more sense?

File: controllers/buildpipeline/buildpipeline_controller_test.go (hunk 7 above)
Comment (hongweiliu17) [reply controllers/buildpipeline/buildpipeline_controller_test.go:49]: Yes, thanks!

File: loader/integration.go (hunk 8 above)
Comment (hongweiliu17) [reply loader/integration.go:66]: Because `prepareSnapshot` calls function `GetAllApplicationComponents` from `loader` package, if I add `prepareSnapshot` to `gitops`  need to import `loader`, I am afraid we will have import conflict. But let me try it. Thanks.

File: controllers/buildpipeline/buildpipeline_controller.go (hunk 5 above)
Comment (hongweiliu17) [reply controllers/buildpipeline/buildpipeline_controller.go]: Both of the two cache functions should be removed.

File: loader/integration.go (hunk 8 above)
Comment (dirgim) [reply loader/integration.go:66]: If that doesn't work, we can refactor the `PrepareSnapshot()` to take a list of ApplicationComponents instead of [fetching them within the function](https://github.com/redhat-appstudio/integration-service/blob/1a4175c9c8bc633f3489868198982b99fb0342b7/controllers/pipeline/pipeline_adapter.go#L414) - that way the loader function would not be used within it, but from the adapters, wdyt?

Comment (hongweiliu17) [reply loader/integration.go:66]: This is a good idea.

Comment (hongweiliu17) [reply loader/integration.go:66]: Maybe I can also change [integrationPipelineRuns](https://github.com/redhat-appstudio/integration-service/blob/1a4175c9c8bc633f3489868198982b99fb0342b7/loader/integration.go#L33) as a parameter and move [GetLatestPipelineRunForSnapshotAndScenario()](https://github.com/redhat-appstudio/integration-service/blob/1a4175c9c8bc633f3489868198982b99fb0342b7/loader/integration.go#L32) out of `loader` and back to the older package. Last time, due to the same reason, I move it to `loader` package?  @dirgim

Comment (dirgim) [reply loader/integration.go:66]: That sounds reasonable to me.

Comment (dirgim) [reply loader/integration.go:66]: It is on the edge of expanding the scope for this PR, but since it's a function that's called from the [current pipeline adapter](https://github.com/redhat-appstudio/integration-service/blob/1a4175c9c8bc633f3489868198982b99fb0342b7/controllers/pipeline/pipeline_adapter.go#L390) which is being refactored, I think it can be included.

Comment (hongweiliu17) [reply loader/integration.go:66]: Thanks. So let me move [GetLatestPipelineRunForSnapshotAndScenario()](https://github.com/redhat-appstudio/integration-service/blob/1a4175c9c8bc633f3489868198982b99fb0342b7/loader/integration.go#L32) in a new PR, otherwise this PR will be bigger and bigger and not easy to review.

Comment (hongweiliu17) [reply loader/integration.go]: `PrepareSnapshot()` is moved to the snapshot.go under `gitops` pkg, thanks!

File: controllers/integrationpipeline/integrationpipeline_controller.go (hunk 3 above)
Comment (hongweiliu17) [reply controllers/integrationpipeline/integrationpipeline_controller.go]: Thanks. I guessed so. `Git` is so smart.

**Review by MartinBasti — APPROVED** [review]
LGTM

**Review by dirgim — APPROVED** [review]
LGTM, nice work

File: controllers/integrationpipeline/integrationpipeline_controller.go [hunk 9]
```diff
@@ -99,19 +93,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
-		}
-	} else if pipelineType == tekton.PipelineRunTestType {
-		application, err = loader.GetApplicationFromPipelineRun(r.Client, ctx, pipelineRun)
-		if err != nil {
-			logger.Error(err, "Failed to get Application from the pipelineRun")
-			return ctrl.Result{}, err
-		}
+	application, err = loader.GetApplicationFromPipelineRun(r.Client, ctx, pipelineRun)
+	if err != nil {
+		logger.Error(err, "Failed to get Application from the test pipelineRun",
```
Comment (dheerajodha) [inline controllers/integrationpipeline/integrationpipeline_controller.go:98]:
```suggestion
		logger.Error(err, "Failed to get Application from the integration pipelineRun",
```

File: controllers/integrationpipeline/integrationpipeline_controller_test.go [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+				},
+			}
+		})
+
+		AfterEach(func() {
+			err := k8sClient.Delete(ctx, integrationPipelineRunNoComponent)
+			Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+		})
+
+		It("reconcile with application taken from pipelinerun (test pipeline)", func() {
```
Comment (dheerajodha) [inline controllers/integrationpipeline/integrationpipeline_controller_test.go:310]:
```suggestion
		It("reconcile with application taken from pipelinerun (integration pipeline)", func() {
```

File: helpers/integration_test.go [hunk 11]
```diff
@@ -29,22 +29,22 @@ var _ = Describe("Pipeline Adapter", Ordered, func() {
 
 	var (
 		//two test pipeline for integrationTestScenario
```
Comment (dheerajodha) [inline helpers/integration_test.go:31]:
```suggestion
		//two integration pipeline for integrationTestScenario
```

File: controllers/integrationpipeline/integrationpipeline_adapter.go [hunk 12]
```diff
@@ -33,11 +30,10 @@ import (
 	"github.com/redhat-appstudio/integration-service/tekton"
 	"github.com/redhat-appstudio/operator-toolkit/controller"
 	tektonv1beta1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
-	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Adapter holds the objects needed to reconcile a Pipeline.
+// Adapter holds the objects needed to reconcile a integration PipelineRun.
```
Comment (dheerajodha) [inline controllers/integrationpipeline/integrationpipeline_adapter.go:36]:
```suggestion
// Adapter holds the objects needed to reconcile a integration PipelineRun.
```
```suggestion
// Adapter holds the objects needed to reconcile an integration PipelineRun.
```

File: controllers/integrationpipeline/integrationpipeline_adapter.go [hunk 13]
```diff
@@ -408,108 +282,15 @@ func (a *Adapter) getAllPipelineRunsForSnapshot(snapshot *applicationapiv1alpha1
 	return &integrationPipelineRuns, nil
 }
 
-// prepareSnapshot prepares the Snapshot for a given application and the updated component (if any).
-// In case the Snapshot can't be created, an error will be returned.
-func (a *Adapter) prepareSnapshot(application *applicationapiv1alpha1.Application, component *applicationapiv1alpha1.Component, newContainerImage string, newComponentSource *applicationapiv1alpha1.ComponentSource) (*applicationapiv1alpha1.Snapshot, error) {
+// prepareCompositeSnapshot prepares the Snapshot for a given application,
```
Comment (dheerajodha) [inline controllers/integrationpipeline/integrationpipeline_adapter.go:285]: Optional
```suggestion
// prepareCompositeSnapshot prepares the Composite Snapshot for a given application,
```

File: controllers/integrationpipeline/integrationpipeline_controller.go (hunk 9 above)
Comment (dheerajodha) [reply controllers/integrationpipeline/integrationpipeline_controller.go:98]: We decided to address this in a separate PR.

File: controllers/integrationpipeline/integrationpipeline_controller_test.go (hunk 10 above)
Comment (dheerajodha) [reply controllers/integrationpipeline/integrationpipeline_controller_test.go:310]: We decided to address this in a separate PR.

File: helpers/integration_test.go (hunk 11 above)
Comment (dheerajodha) [reply helpers/integration_test.go:31]: We decided to address this in a separate PR.

File: controllers/integrationpipeline/integrationpipeline_adapter.go (hunk 12 above)
Comment (dheerajodha) [reply controllers/integrationpipeline/integrationpipeline_adapter.go:36]: We decided to address this in a separate PR.

File: controllers/integrationpipeline/integrationpipeline_adapter.go (hunk 13 above)
Comment (dheerajodha) [reply controllers/integrationpipeline/integrationpipeline_adapter.go:285]: We decided to address this in a separate PR.

**Comment by dirgim** [comment]
Merging this after getting agreement from the team.
